### ISSUE TITLE
Support recursive generics

### ISSUE THREAD

**Issue by marcob** [immutables/immutables#418]
```java
public interface MyBase {

  int getId();
}

public interface MyContent {}

public interface IHaveWithIdMethod<T extends MyBase> {

  T withId(int id); // I that having getId in MyBase will generate this but I need an interface to use commonly on multiple class that inherit from MyBase
}

public interface IHaveGetContentAndWithId<T extends MyContent, A extends MyBase> extends MyBase, IHaveWithIdMethod<A> {

  T getContent();
}

@Immutable
@HubSpotStyle
public interface AbstractSomething extends IHaveGetContentAndWithId<MySomethingContent, Something> {

  // this override is not needed with 2.2, hooray! \o/
  //  @Override
  //  MySomethingContent getContent();

  // without this override it generates the withId method that returns T instead of Something for the Json class ;(
  @Override
  Something withId(int id);
}
```

**Comment by elucash** [comment]
Thank you for reporting this. It's not about recursive generics, it's about not-yet-generated types and javac bugs. I've made a workaround so this case might compile without `Something withId(int id)`. But if this example would be more complicated I doubt it can be resolvable.

Just a note, if you use quite complex combination of subtyping, generics and not-yet-generated types, I would recommend to switch to the style which tries to not expose not-yet-generated types directly as illustrated at the top of a [README.md](https://github.com/immutables/immutables/blob/master/README.md)

**Comment by elucash** [comment]
released in 2.2.12
